src/ucp/api/ucp.h

```diff
@@ -679,10 +682,40 @@ typedef struct ucp_ep_params {
      * Handler to process transport level failure.
      */
     ucp_err_handler_t       err_handler;
+
+    /**
+     * User data associated with an endpoint. See @ref ucp_ep_poll_t.
+     * @a user_data redefines @ref ucp_err_handler_t::arg if set both
+     */
+    void                    *user_data;
 } ucp_ep_params_t;
 
 
 /**
+ * @ingroup UCP_ENDPOINT
+ * @brief Output parameter of @ref ucp_stream_worker_poll function.
+ *
+ * The structure defines the endpoint and its user data.
+ */
+typedef struct {
+    /**
+     * Endpoint handle.
+     */
+    ucp_ep_h    ep;
+
+    /**
+     * User data associated with an endpoint passed in
+     * @ref ucp_ep_params_t::user_data.
+     */
+    void        *user_data;
+
+    /**
+     * Reserved for future use.
+     */
+    unsigned    flags;
+} ucp_stream_poll_ep_t;
```

Comment:
we pass sizeof() as an additional parameter to the poll function, but this will add more overhead
if compatibility is important i would suggest to add some reserved space (e.g char reserved[16])